fluss-server/src/main/java/org/apache/fluss/server/zk/data/lake/LakeTableHelper.java

```diff
@@ -44,99 +46,96 @@ public LakeTableHelper(ZooKeeperClient zkClient, String remoteDataDir) {
+                            new TableBucketOffsets(
+                                    tableId, lakeTableSnapshot.getBucketLogEndOffset()));
+            lakeTableSnapshot = new LakeTableSnapshot(tableId, tableBucketOffsets.getOffsets());
         }
+        zkClient.upsertLakeTable(
+                tableId, new LakeTable(lakeTableSnapshot), optPreviousLakeTable.isPresent());
+    }
 
-        // store the lake table snapshot into a file
-        FsPath lakeTableSnapshotFsPath =
-                storeLakeTableSnapshot(tableId, tablePath, lakeTableSnapshot);
-
-        LakeTable.LakeSnapshotMetadata lakeSnapshotMetadata =
-                new LakeTable.LakeSnapshotMetadata(
-                        lakeTableSnapshot.getSnapshotId(),
-                        // use the lake table snapshot file as the tiered offsets file since
-                        // the table snapshot file will contain the tiered log end offsets
-                        lakeTableSnapshotFsPath,
-                        // currently, readableOffsetsFilePath is always same with
-                        // tieredOffsetsFilePath, but in the future we'll commit a readable offsets
-                        // separately to mark what the readable offsets are for a snapshot since
-                        // in paimon dv table, tiered log end offsets is not same with readable
-                        // offsets
-                        lakeTableSnapshotFsPath);
-
-        // currently, we keep only one lake snapshot metadata in zk,
-        // todo: in solve paimon dv union read issue #2121, we'll keep multiple lake snapshot
-        // metadata
+    public void addLakeTableSnapshotMetadata(
+            long tableId, LakeTable.LakeSnapshotMetadata lakeSnapshotMetadata) throws Exception {
+        Optional<LakeTable> optPreviousLakeTable = zkClient.getLakeTable(tableId);
+        List<LakeTable.LakeSnapshotMetadata> previousLakeSnapshotMetadatas = null;
+        if (optPreviousLakeTable.isPresent()) {
+            previousLakeSnapshotMetadatas = optPreviousLakeTable.get().getLakeSnapshotMetadatas();
+        }
         LakeTable lakeTable = new LakeTable(lakeSnapshotMetadata);
         try {
             zkClient.upsertLakeTable(tableId, lakeTable, optPreviousLakeTable.isPresent());
         } catch (Exception e) {
             LOG.warn("Failed to upsert lake table snapshot to zk.", e);
-            // discard the new lake snapshot metadata
-            lakeSnapshotMetadata.discard();
             throw e;
         }
 
+        // currently, we keep only one lake snapshot metadata in zk,
+        // todo: in solve paimon dv union read issue #2121, we'll keep multiple lake snapshot
+        // metadata
+        // discard previous lake snapshot metadata
+        if (previousLakeSnapshotMetadatas != null) {
+            previousLakeSnapshotMetadatas.forEach(LakeTable.LakeSnapshotMetadata::discard);
+        }
+    }
+
+    public TableBucketOffsets upsertTableBucketOffsets(
```

Comment:
It sounds like this method will `upsert` bucket offsets into zk (very similar to other methods like `upsertLakeTable`). However, it doesn't.  Would be better to move the logic into `org.apache.fluss.server.RpcServiceBase#prepareCommitLakeTableSnapshot`.